lib/Logger.js

```diff
@@ -43,6 +45,26 @@ module.exports = class Logger {
     this._output({ level, record });
   }
 
+  drain(timeout) {
+    this._draining =
+      this._draining ||
+      new Promise((resolve, reject) => {
+        const timerId =
+          timeout &&
+          setTimeout(() => {
+            reject(new Error("Timedout waiting for logger to drain"));
+          });
+        const waitUntilDrained = () => {
+          if (this._promises.size > 0) return setImmediate(waitUntilDrained);
```

Comment:
I am concerned this may loop quite tightly while there are pending writes, however there will still only be one draining function on the event loop at any one time, and I believe the IO queue is checked first so the setImmediate shouldn't block anything. That said, I suspect there will be a CPU spike, so it may be better to use a setInterval with a low timeout value (e.g. 50ms)

Reply:
A quick test reveals setImmediate takes 30% of CPU compared to setTimeout with a 10ms delay was only 1%. I'll update to setTimeout.